x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/AbstractFallibleArrayState.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.compute.aggregation;
+
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.common.util.BitArray;
+import org.elasticsearch.core.Releasable;
+import org.elasticsearch.core.Releasables;
+
+public class AbstractFallibleArrayState implements Releasable {
+    protected final BigArrays bigArrays;
+
+    private BitArray seen;
+    private BitArray failed;
```

Comment:
I think it's worth a comment here about when these are initialized. And I think it's also worth a comment about why you didn't use an `enum` or an array of two-bit values. I think it's correct to use to BitArrays here because frequently you won't need either. And when you need `seen` you often won't need `failed`. That's why I think you did it this way. And, also, it's probably easier to do it this way....

Reply:
Yeah, we could do it with an enum-like array. We would lose tho the "false by default" feature of those bitarrays that avoid growing them for falses, and it would take more memory I expect.
Tbh, I didn't consider it, this looked quite minimal.

Going to comment both of them. If you think using other kind of container would be worth it, just tell me!